TELCODOCS-1188 creating new PR about adding new parameter (#62354)
openshift/openshift-docs: modules/nw-sriov-networknodepolicy-object.adoc

  mtu: <mtu> <6>
  needVhostNet: false <7>
  numVfs: <num> <8>

>   externallyManage: false <9>

Reviewer: externallyManage -> externallyManaged

<8> The number of the virtual functions (VF) to create for the SR-IOV physical network device. For an Intel network interface controller (NIC), the number of VFs cannot be larger than the total VFs supported by the device. For a Mellanox NIC, the number of VFs cannot be larger than `128`.

> <9> Set `externallyManage` to `true` to ensure some virtual functions (VFs) are reserved for host managed systems. With the value set to `false` the SR-IOV Network Operator manages and configures all allocated VFs.

Reviewer: maybe something like 

```
Allow the SR-IOV network operator to configure and allocate a subset of virtual functions from
a physical function that is configured externally from SR-IOV network operator.
```

please take a look on the design proposal in the sriov repo https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/476/files

<8> The number of the virtual functions (VF) to create for the SR-IOV physical network device. For an Intel network interface controller (NIC), the number of VFs cannot be larger than the total VFs supported by the device. For a Mellanox NIC, the number of VFs cannot be larger than `128`.

<9> Set `externallyManaged` to `true` to allow the SR-IOV Network Operator to use all or a subset of externally managed virtual functions (VFs) and attach them to pods. With the value set to `false` the SR-IOV Network Operator manages and configures all allocated VFs.
+
[NOTE]
====
When `externallyManaged` is set to `true`, you must create the Virtual Functions (VFs) before applying the policy. If not, the webhook will block the request.

> If `externallyManaged` is set to `false`, the SR-IOV Network Operator handles the creation and management of VFs, including resetting them if necessary. Therefore to use VFs on the host system they must be created manually and `externallyManaged` must be set to `true` so the SR-IOV Operator Network will not take any actions on the created VFs.

Reviewer: so the SR-IOV Network Operator will not take any actions on the PF  and the VFs that are not in the policy nitSelector.

Author: Hi @SchSeba is this feature now released. I was waiting for the z-stream. Is it out and in what z-stream ?

Author: HI @SchSeba should this read "so the SR-IOV Network Operator will not take any actions on the PF and the VFs that are not in the policy nodeSelector" ?

Author: @SchSeba got it I think "so the SR-IOV Network Operator will not take any actions on the PF and the VFs that are not defined in the policy nicSelector"

Reviewer: yep I think it sounds good thanks!

the backport to 4.14.z is WIP should be done this week

<8> The number of the virtual functions (VF) to create for the SR-IOV physical network device. For an Intel network interface controller (NIC), the number of VFs cannot be larger than the total VFs supported by the device. For a Mellanox NIC, the number of VFs cannot be larger than `128`.

> <9> Set `externallyManaged` to `true` to allow the SR-IOV Network Operator to configure and allocate a subset of virtual functions from

Reviewer 2: The first sentence says the Operator configures a subset of VFs when `externallyManaged=true`. The second sentence says that the Operator configures the VFs when `externallyManaged=false`.

When `externallyManaged=true`, the Operator will not allocate or configure the VFs. The Operator can use the VFs to attach them to Pods but not manage them (i.e. create/allocate and configure)

Author: I thought the key point was that with this set to true the SR-IOV Network Operator is can configure a subset of "all" VF whereas with it set to false (think default) the SR-IOV Network Operator does not have that granular control and configures "all"

Reviewer 2: On the contrary. The parameter is called `externallyManaged` which hints that VFs are not to be managed by the Operator. The default value is `false`, yes.

Reviewer 2: Please consider the following suggestion:

"Set `externallyManaged` to `true` to allow the SR-IOV Network Operator to use all or a subset of externally managed virtual functions and attach them to Pods. With the value set to `false` the SR-IOV Network Operator manages and configures all allocated VFs."

Author: Thank you, @cgoncalves "to use **all** or a subset of externally managed virtual functions" seems like a important distinction which I must have missed.

Reviewer 3: @cgoncalves  You are absolutely right:
* If externallyCreated is false (default) - SR-IOV must create/recreate the VFs on its own.
* if externallyCreated is true - SR-IOV should use VFs created earlier.

Reviewer 3: @kquinn1204
 BTW This line is only relevant if externallyCreated is true. 

> Before applying the policy, it is necessary to create the VFs first; otherwise, the webhook will block the request.

Author: @evgenLevin I made that clarification thanks

> a physical function that is configured externally by using the SR-IOV Network Operator. With the value set to `false` the SR-IOV Network Operator manages and configures all allocated VFs. test

Reviewer 3: Please remove "test"

Author: yep I just added that to try to trigger rebuild of preview. Will remove Otherwise is it OK ?

Author: just removed. Thanks

Author: OK see LGTM now thanks @evgenLevin

<8> The number of the virtual functions (VF) to create for the SR-IOV physical network device. For an Intel network interface controller (NIC), the number of VFs cannot be larger than the total VFs supported by the device. For a Mellanox NIC, the number of VFs cannot be larger than `128`.

<9> Set `externallyManaged` to `true` to allow the SR-IOV Network Operator to use all or a subset of externally managed virtual functions (VFs) and attach them to pods. With the value set to `false` the SR-IOV Network Operator manages and configures all allocated VFs.
+
[NOTE]
====

> When `externallyManaged` is set to `true`, you must create the Virtual Functions (VFs) before applying the policy. If not, the webhook will block the request.

Reviewer 4: Which VFs, the ones used by the Host?
And which policy, please clarify we are specifically referring to the SriovNetworkNodePolicy resource
And finally, which webhook are we talking about and which request will it block?

> If `externallyManaged` is set to `false`, the SR-IOV Network Operator handles the creation and management of VFs, including resetting them if necessary. Therefore to use VFs on the host system they must be created manually and `externallyManaged` must be set to `true` so the SR-IOV Network Operator will not take any actions on the PF and the VFs that are not defined in the policy `nicSelector`.

Reviewer 4: What does 'resetting` mean in this context?

Suggest to clarify that any VFs used on the host system must be created through NMState (not manually), while the SR-IOV operator will manage VFs presented as Pod secondary interfaces as always

====

> <10> The NIC selector identifies the device for the Operator to configure. You do not have to specify values for all the parameters. It is recommended to identify the network device with enough precision to avoid selecting a device unintentionally.

Reviewer 4: ```suggestion
<10> The NIC selector identifies the device for which this resource applies to. You do not have to specify values for all the parameters. It is recommended to identify the network device with enough precision to avoid selecting a device unintentionally.
```

+
If you specify `rootDevices`, you must also specify a value for `vendor`, `deviceID`, or `pfNames`. If you specify both `pfNames` and `rootDevices` at the same time, ensure that they refer to the same device. If you specify a value for `netFilter`, then you do not need to specify any other parameter because a network ID is unique.

> <11> Optional: The vendor hexadecimal code of the SR-IOV network device. The only allowed values are `8086` and `15b3`.

Reviewer 4: ```suggestion
<11> Optional: The vendor hexadecimal vendor identifier of the SR-IOV network device. The only allowed values are `8086` (Intel) and `15b3` (Mellanox).
```

> <12> Optional: The device hexadecimal code of the SR-IOV network device. For example, `101b` is the device ID for a Mellanox ConnectX-6 device.

Reviewer 4: ```suggestion
<12> Optional: The device hexadecimal device identifier of the SR-IOV network device. For example, `101b` is the device ID for a Mellanox ConnectX-6 device.
```

> <13> Optional: An array of one or more physical function (PF) names for the device.

Reviewer 4: ```suggestion
<13> Optional: An array of one or more physical function (PF) names the resource should apply to.
```

> <14> Optional: An array of one or more PCI bus addresses for the PF of the device. Provide the address in the following format: `0000:02:00.1`.

Reviewer 4: ```suggestion
<14> Optional: An array of one or more PCI bus addresses the resource should apply to.  
Example: `0000:02:00.1`.
```
A PF can not be spread across multiple PCI devices

> <15> Optional: The platform-specific network filter. The only supported platform is {rh-openstack-first}. Acceptable values use the following format: `openstack/NetworkID:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx`. Replace `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` with the value from the `/var/config/openstack/latest/network_data.json` metadata file.

Reviewer 4: Please clarify (or link) what a network filter is

> <16> Optional: The driver type for the virtual functions. The only allowed values are `netdevice` and `vfio-pci`. The default value is `netdevice`.

Reviewer 4: ```suggestion
<16> Optional: The driver to configure for the VFs created from this resource. The only allowed values are `netdevice` and `vfio-pci`. The default value is `netdevice`.
```

+
For a Mellanox NIC to work in DPDK mode on bare metal nodes, use the `netdevice` driver type and set `isRdma` to `true`.

<17> Optional: Configures whether to enable remote direct memory access (RDMA) mode. The default value is `false`.
+
If the `isRdma` parameter is set to `true`, you can continue to use the RDMA-enabled VF as a normal network device. A device can be used in either mode.
+
Set `isRdma` to `true` and additionally set `needVhostNet` to `true` to configure a Mellanox NIC for use with Fast Datapath DPDK applications.

<18> Optional: The link type for the VFs. The default value is `eth` for Ethernet. Change this value to 'ib' for InfiniBand.
+
When `linkType` is set to `ib`, `isRdma` is automatically set to `true` by the SR-IOV Network Operator webhook. When `linkType` is set to `ib`, `deviceType` should not be set to `vfio-pci`.
+
Do not set linkType to `eth` for SriovNetworkNodePolicy, because this can lead to an incorrect number of available devices reported by the device plugin.

> <19> Optional: To enable hardware offloading, the `eSwitchMode` field must be set to `"switchdev"`.

Reviewer 4: Please specify what type of hardware offloading is enabled

Author: Good question I did not directly write this piece do you know what type of hardware offloading is enabled ?

Reviewer 4: Not really and since this is not clear to me, I'm guessing it won't be clear to a customer either
